creator-node/src/URSMService.js

```diff
@@ -0,0 +1,234 @@
+    this.delegatePrivateKey = nodeConfig.get('delegatePrivateKey')
+    this.spOwnerWallet = nodeConfig.get('spOwnerWallet')
+
+    // Requires properly initialized libs instance and non-zero spID
+    // Currently spID is configured inside snapbackSM
+    if (!this.audiusLibs || !this.spID || !this.delegateOwnerWallet || !this.delegatePrivateKey || !this.spOwnerWallet) {
+      throw new Error('nope')
+    }
+
+    this.numSignaturesRequired = 3
+
+    this.logInfo(`URSMSERVICE CONSTRUCTED`)
+  }
+
+  logInfo (msg) {
+    logger.info(`URSMService || ${msg}`)
+  }
+
+  logError (msg) {
+    logger.error(`URSMService ERROR || ${msg}`)
+  }
+
+  /**
+   * TODO DESCRIPTION
+   *
+   * Steps: 
+   *  1. Fetch node record from L1 ServiceProviderFactory for spID
+   *    a. Short-circuit if no L1 record found
+   *  2. Fetch node record from L2 UserReplicaSetManager for spID
+   *    a. Short-circuit if L2 record for node already matches L1 record (i.e. delegateOwnerWallets match)
+   *  3. Fetch list of all nodes registered on URSM, in order to submit requests for proposal
+   *    a. Randomize list to minimize bias
+   *    b. Organize list to ensure requests are sent to nodes with unique ownerWallets
+   *  4. Submit requests for proposal to nodes until 3 successful signatures received
+   *    a. Must have signatures from nodes with unique ownerWallets
+   *    b. Error if all available nodes contacted without 3 successful signatures
+   *  5. Submit registration transaction to URSM with signatures
+   */
+  async init () {
+    this.logInfo('STARTING INIT')
+
+    /**
+     * 1. Fetch node record from L1 ServiceProviderFactory for spID
+     */
+    const spRecordFromSPFactory = await this.audiusLibs.ethContracts.ServiceProviderFactoryClient.getServiceEndpointInfo(
+      'content-node',
+      this.spID
+    )
+    let {
+      owner: ownerWalletFromSPFactory,
+      delegateOwnerWallet: delegateOwnerWalletFromSPFactory,
+      endpoint: endpointFromSPFactory
+    } = spRecordFromSPFactory
+    delegateOwnerWalletFromSPFactory = delegateOwnerWalletFromSPFactory.toLowerCase()
+
+
+    /**
+     * 1-a. Short-circuit if no L1 record found
+     */
+    if (
+      LibsUtils.isZeroAddress(ownerWalletFromSPFactory)
+      || LibsUtils.isZeroAddress(delegateOwnerWalletFromSPFactory)
+      || !endpointFromSPFactory
+    ) {
+      throw new Error('L1recordbad')
+    }
+
+    /**
+     * 2. Fetch node record from L2 UserReplicaSetManager for spID
+     */
+    const delegateOwnerWalletFromURSM = (
+      (await this.audiusLibs.contracts.UserReplicaSetManagerClient.getContentNodeWallets(this.spID))
+      .delegateOwnerWallet
+    ).toLowerCase()
+
+    /**
+     * 2-a. Short-circuit if L2 record for node already matches L1 record (i.e. delegateOwnerWallets match)
+     */
+    if (delegateOwnerWalletFromSPFactory === delegateOwnerWalletFromURSM) {
+      throw new Error('L2recordmatch')
```

Comment:
verify ownerWallet on chain for this endpoint,delegateOwnerWallet matches the configured ownerWallet parameter in this content node (spOwnerWallet i believe)